### PR TITLE
MV3 click to load support

### PR DESCRIPTION
**Pull request by ladamski**
**Reviewer:**

## Description:

Add support for click to load in MV3
https://app.asana.com/0/72649045549333/1202839073279061/f
supporting PRs:
https://github.com/duckduckgo/content-scope-scripts/pull/189
https://github.com/duckduckgo/ddg2dnr/pull/37

## Automated tests:
- [ ] Unit tests
- [ ] Integration tests

###### Reviewer Checklist:
- [ ] **Ensure the PR solves the problem**
- [ ] **Review every line of code**
- [ ] **Ensure the PR does no harm by testing the changes thoroughly**
- [ ] **Get help if you're uncomfortable with any of the above!**
- [ ] Determine if there are any quick wins that improve the implementation

###### PR Author Checklist:
- [ ] Get advice or leverage existing code
- [ ] Agree on technical approach with reviewer (if the changes are nuanced)
- [ ] Ensure that there is a testing strategy (and documented non-automated tests)
- [ ] Ensure there is a documented monitoring strategy (if necessary)
- [ ] Consider systems implications

### CODE REVIEW

File: shared/js/background/before-request.es6.js
Comment (kzar) [inline shared/js/background/before-request.es6.js:100]: I'm suprised ESLint didn't catch this, perhaps the GitHub action isn't working?

File: shared/js/background/before-request.es6.js
Comment (kzar) [inline shared/js/background/before-request.es6.js:106]: I think an empty array is considered truthy in JavaScript. How about this?
```suggestion
                if (customRulesIds.length === 0) {
```

File: shared/js/background/before-request.es6.js
Comment (kzar) [inline shared/js/background/before-request.es6.js:101]: Would be nice to explain the idea in this comment here a bit more, even if you only give CTL rules as an example.

File: shared/js/background/before-request.es6.js
Comment (kzar) [inline shared/js/background/before-request.es6.js:107]:
```suggestion
                    chrome.declarativeNetRequest.updateSessionRules({ removeRuleIds: customRulesIds })
```

File: shared/js/background/before-request.es6.js
Comment (kzar) [inline shared/js/background/before-request.es6.js:102]: IMO let's move this logic into your `custom-rules-manager.js` module as a function and import+call it here.

File: shared/js/background/classes/custom-rules-manager.js
Comment (kzar) [inline shared/js/background/classes/custom-rules-manager.js:14]: Please can you add a JSDoc comment for this? Then as a bonus tslint will check the input types are correct.

File: shared/js/background/classes/tab-state.js
Comment (kzar) [inline shared/js/background/classes/tab-state.js:35]: Could you specify this type in more detail? So we avoid typos with the property names?

File: shared/js/background/declarative-net-request.js
Comment (kzar) [inline shared/js/background/declarative-net-request.js:338]:
```suggestion
            const ruleIds = rules.map(({ id }) => id).filter(isValidSessionId)
```

File: shared/js/background/declarative-net-request.js
Comment (kzar) [inline shared/js/background/declarative-net-request.js:337]: Maybe remove this comment? Or if not, we should at least move this below where we define `ruleIds`.

File: shared/js/background/declarative-net-request.js
Comment (kzar) [inline shared/js/background/declarative-net-request.js:334]: IMO let's move this into `dnr-session-rule-id.js`. Perhaps we could rename stuff in that module a bit to make it clear the distinction between session rules with incremented ID and other session rules too?

File: shared/js/background/declarative-net-request.js
Comment (kzar) [inline shared/js/background/declarative-net-request.js:335]: Nit: Might as well use `await` here, since the function is `async` anyway.

File: shared/js/background/declarative-net-request.js
Comment (kzar) [inline shared/js/background/declarative-net-request.js:339]: Nit: Might as well `await` here too, so that the returned Promise only resolves when this is done. So far, it will resolve before then which could be a foot-gun in the future.

File: shared/js/background/dnr-session-rule-id.js
Comment (kzar) [inline shared/js/background/dnr-session-rule-id.js:25]: Nice, thanks for fixing some typos :1st_place_medal:

File: shared/js/background/events.es6.js
Comment (kzar) [inline shared/js/background/events.es6.js:72]: Perhaps expand this comment, or better we could add a JSDoc comment for the function. Either way, we should explain the real idea of this somewhere.  (e.g. "We increment the rule IDs for some session rules, starting at STARTING_RULE_ID and keep a note of the next rule ID in session storage. Sometimes (e.g. on extension update) session storage can be cleared, while session rules are not. To avoid an inconsistent state, let's clear session rules on extension update too.")

File: shared/js/content-scripts/content-scope-messaging.js
Comment (kzar) [inline shared/js/content-scripts/content-scope-messaging.js:17]:
```suggestion
        if (!detail) {
```

File: shared/js/content-scripts/content-scope-messaging.js
Comment (kzar) [inline shared/js/content-scripts/content-scope-messaging.js:13]: Is this message coming from our main_world script? If so, couldn't the website spoof such a message too? Perhaps we could use the secret in an event name to mitigate?

File: shared/js/content-scripts/content-scope-messaging.js
Comment (kzar) [inline shared/js/content-scripts/content-scope-messaging.js:23]: Doesn't look like `stringifiedArgs` is being used anywhere?

File: shared/js/background/before-request.es6.js
Comment (ladamski) [reply shared/js/background/before-request.es6.js]: 👍  Went through and cleaned up some lint errors in general

Comment (ladamski) [reply shared/js/background/before-request.es6.js]: This seems to work?  Good idea on the comment either way, esp if we run into issues later.

File: shared/js/background/before-request.es6.js
Comment (ladamski) [reply shared/js/background/before-request.es6.js:101]: 👍

File: shared/js/background/dnr-session-rule-id.js
Comment (ladamski) [reply shared/js/background/dnr-session-rule-id.js:24]: 👍

File: shared/js/background/before-request.es6.js
Comment (ladamski) [reply shared/js/background/before-request.es6.js]: Good catch! Fixed a few of these but missed this one

File: shared/js/background/before-request.es6.js
Comment (ladamski) [reply shared/js/background/before-request.es6.js:105]: 👍

File: shared/js/background/declarative-net-request.js
Comment (ladamski) [reply shared/js/background/declarative-net-request.js:334]: Not sure I follow, `dnr-session-rule-id.js` seems to be focused on managing rule IDs, while `declarative-net-request.js` is about managing the rules themselves (eg. `toggleUserAllowlistDomain`) which is what `flushSessionRules` does?

File: shared/js/background/declarative-net-request.js
Comment (ladamski) [reply shared/js/background/declarative-net-request.js:335]: That is handled by the caller (which is the right pattern IMHO, let the caller decide whether to await or resolve as promise)

File: shared/js/content-scripts/content-scope-messaging.js
Comment (ladamski) [reply shared/js/content-scripts/content-scope-messaging.js:13]: No the communication from the main world content scripts hasn't changed, it just required some additional proxying internally for MV3

File: shared/js/background/classes/tab-state.js
Comment (ladamski) [reply shared/js/background/classes/tab-state.js:35]: I expanded it a bit but the keys are dynamically generated

File: shared/js/background/declarative-net-request.js
Comment (kzar) [reply shared/js/background/declarative-net-request.js:346]: The caller will be returned a Promise either way, since this is an async function. The caller can wait (or not) for that Promise to resolve either way. My suggestion would just mean that the returned Promise would resolve to the result (if any) of updateSessionRules, instead of resolving to a Promise that in turn resolves to a Promise which in turn resolves to the result on updateSessionRules.

Comment (ladamski) [reply shared/js/background/declarative-net-request.js:346]: Oops you're right, insofar I forgot to remove the `async`.  :)

I noticed we seem to have fallen a bit into the pattern of using `await` just for return values, which in turn requires the containing function itself to be async, which is a NOOP in practice and creates cognitive overhead.   `refreshUserAllowlistRules()` right above is an example of this.

Maybe this better illustrates my point?

```
async function go() {
    function a1() {
        return Promise.resolve(true)
    }

    function a2() {
        return a1()
    }

    function a3() {
        return a2()
    }

    console.log('go', a3())
    console.log('await go ', await a3())
}

async function gonuts() {
    async function a1() {
        return await Promise.resolve(true)
    }

    async function a2() {
        return await a1()
    }

    async function a3() {
        return await a2()
    }

    console.log('gonuts', a3())
    console.log('await gonuts', await a3())
}

go()
gonuts()
```
results in
```
go Promise { true }
gonuts Promise { <pending> }
await go  true
await gonuts true
```

So I think if we see patterns where we're just doing `async .... return await x()`, that should give us pause?

**Comment by ladamski** [comment]
Ok should be good to go here (modulo merging https://github.com/duckduckgo/content-scope-scripts/pull/189 and updating dependencies accordingly).  

Local int & int-mv3 tests passing too.

**Comment by ladamski** [comment]
Figured out the issue with the integration test (hopefully), the content scope script dependency was just pointing at the branch name which was cached, so I updated to the latest commit.  Once that branch is merged, we'll need to update this one to that release before merging.

**Comment by ladamski** [comment]
Ok I spoke too soon.  A simple timeout obscured the error, but its still there.   Really not sure what's going on, but please try it locally and see if it reproduces?  Still passing consistently locally for me.

File: shared/js/content-scripts/content-scope-messaging.js
Comment (kzar) [reply shared/js/content-scripts/content-scope-messaging.js:13]: OK, so long as you're confident that websites can't just throw a "sendMessage" event on window to trick this logic then it sounds good to me.

File: shared/js/background/events.es6.js
Comment (kzar) [inline shared/js/background/events.es6.js:54]: Have you filed an issue for this? Cc @jdorweiler

File: shared/js/background/declarative-net-request.js
Comment (kzar) [inline shared/js/background/declarative-net-request.js:484]: Think we should check `ruleIds.length > 0` after filtering instead.

**Review by kzar — APPROVED** [review]
LGTM, but would be nice to get the integration tests passing.

File: shared/js/content-scripts/content-scope-messaging.js
Comment (kzar) [reply shared/js/content-scripts/content-scope-messaging.js:13]: In fact, could you test that? I am a little concerned this could be a security vulnerability, especially considering even the message type is taken from the event - if it's spoofed that would be bad.

File: shared/js/background/events.es6.js
Comment (jdorweiler) [reply shared/js/background/events.es6.js:54]: @ladamski looks like this is used for email autofill. Can you check that it's still working?

Comment (ladamski) [reply shared/js/background/events.es6.js:54]: We discussed this prior somewhere in the CTL project, trying to find a reference to it (Asana permitting).  It only runs on ext update/install.

Comment (ladamski) [reply shared/js/background/events.es6.js:54]: Regarding sendMessage, I think we're probably due for a bit of a small security pass through the CTL code.  We have multiple messaging channels in different browsers (incl Mac), and they have all changed a bit over time, so I'll file a separate task to test this more thoroughly.

Comment (jdorweiler) [reply shared/js/background/events.es6.js:54]: 👍 for autofill. Can you make a followup to make sure we test it after this.

**Comment by jdorweiler** [comment]
Not sure about the failing test. I can't get it to fail locally.

**Comment by kzar** [comment]
> Not sure about the failing test. I can't get it to fail locally.

Same. I suspect it might be the block list loading slowly, perhaps when we update the bundled block list this will be resolved. Also, Sam's WIP is adding a way to wait for loading a test block list which could help here.

Since we have a deadline here, do you want to disable the Facebook CTL integration tests as well here? I will look into re-enabling them afterwards.

File: shared/js/content-scripts/content-scope-messaging.js
Comment (kzar) [reply shared/js/content-scripts/content-scope-messaging.js:13]: OK, I have tested it and it seems to be a security vulnerability. Try going to https://testpages.kzar.co.uk/insecure-extension-messaging/ and the website tell the extension to open its options page + feedback page.

Comment (kzar) [reply shared/js/content-scripts/content-scope-messaging.js:13]: It turns out that already exists in the released extension, resolving this therefore.

File: shared/js/content-scripts/content-scope-messaging.js
Comment (kzar) [inline shared/js/content-scripts/content-scope-messaging.js:13]: I was wondering why we need to add the listener again here, since it's included in content-scope-scripts. I looked into it, I think that's since we don't have a listener in [content-scope-scripts/inject/chrome-mv3.js](https://github.com/duckduckgo/content-scope-scripts/blob/main/inject/chrome-mv3.js) like we do in [content-scope-scripts/inject/chrome.js](https://github.com/duckduckgo/content-scope-scripts/blob/main/inject/chrome.js). I think it would be better to add this listener to content-scope-scripts instead.

**Review by kzar — APPROVED** [review]

**Pull request merged**
